File: src/skywolf46/NBTUtil/v1_1R1/BukkitVersionUtil.java
package skywolf46.NBTUtil.v1_1R1;

import org.bukkit.Bukkit;
import org.bukkit.Server;

public class BukkitVersionUtil {
    public static String getVersion(Server server) {
        final String packageName = server.getClass().getPackage().getName();

        return packageName.substring(packageName.lastIndexOf('.') + 1);
    }

    public static Class getNMSClass(String className) throws ClassNotFoundException {
        return Class.forName("net.minecraft.server." + getVersion(Bukkit.getServer()) + "." + className);
    }


    public static Class getOBCClass(String className) throws ClassNotFoundException {
        return Class.forName("org.bukkit.craftbukkit." + getVersion(Bukkit.getServer()) + "." + className);
    }
}

File: src/main/kotlin/skywolf46/refnbt/util/BukkitVersionUtil.kt
package skywolf46.refnbt.util

import org.bukkit.Bukkit
import org.bukkit.Server
import skywolf46.extrautility.util.log
import skywolf46.refnbt.util.item.ItemNBTUtil


object BukkitVersionUtil {
    fun getVersion(server: Server): String {
        val packageName = server.javaClass.getPackage().name
        return packageName.substring(packageName.lastIndexOf('.') + 1)
    }

    @Throws(ClassNotFoundException::class)
    fun getNMSClass(className: String): Class<*> {
        return try {
            Class.forName("net.minecraft.server." + getVersion(Bukkit.getServer()) + "." + className)
        } catch (ex: Exception) {
            Void.TYPE
        }
    }

    @Throws(ClassNotFoundException::class)
    fun getOBCClass(className: String): Class<*> {
        return Class.forName("org.bukkit.craftbukkit." + getVersion(Bukkit.getServer()) + "." + className)
    }

    fun checkLegacy() {
        // If lower than 1.8, convert to Legacy mode.
        if (BukkitVersion(1, 8, 9) > current()) {
            log("§e[ReflectedNBTWrapper] §cReflectedNBTWrapper is running as legacy mode")
            ItemNBTUtil.enableLegacy()
        }
    }

    fun current(): BukkitVersion {
        return getVersion(Bukkit.getServer()).replace(Regex("[vR]"), "").split("_").let {
            return@let BukkitVersion(it[0].toInt(), it[1].toInt(), it[2].toInt())
        }
    }
}
